core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java

```diff
@@ -335,12 +344,11 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx,
       localCommand.setTopologyId(command.getTopologyId());
       InvocationFinallyAction handler =
             createLocalInvocationHandler(ch, allFuture, segments, helper, MergingCompletableFuture.moveListItemsToFuture(myOffset));
-      if (retrievals == null) {
+      if (remoteKeys.isEmpty()) {
          invokeNextAndFinally(ctx, localCommand, handler);
       } else {
          // We must wait until all retrievals finish before proceeding with the local command
-         CompletableFuture[] ra = retrievals.toArray(new CompletableFuture[retrievals.size()]);
-         Object result = asyncInvokeNext(ctx, command, CompletableFuture.allOf(ra));
+         Object result = asyncInvokeNext(ctx, command, retrievals);
```

Comment:
@rvansa Ouch, I read lines 348 and 352 invoking the next interceptor and I almost had a heart attack thinking that we're going to have multiple threads accessing the context :)

`CallInterceptor` should be pretty fast, wouldn't it simplify things if we only invoked it after we got all the entries in the context?